sdk/src/main/java/com/mapbox/maps/viewannotation/ViewAnnotationManagerImpl.kt

```diff
@@ -0,0 +1,250 @@
+import com.mapbox.maps.*
+import com.mapbox.maps.plugin.ViewPlugin
+import java.util.concurrent.ConcurrentHashMap
+
+internal class ViewAnnotationManagerImpl(
+  private val mapView: MapView,
+  private val mapboxMap: MapboxMap = mapView.getMapboxMap()
+) : ViewAnnotationManager, ViewAnnotationPositionsListener {
+
+  init {
+    mapView.requestDisallowInterceptTouchEvent(false)
+    mapboxMap.setViewAnnotationPositionsUpdateListener(this)
+  }
+
+  private val asyncInflater by lazy {
+    try {
+      AsyncLayoutInflater(mapView.context)
+    } catch (e: NoClassDefFoundError) {
+      throw RuntimeException(
+        "Please add https://mvnrepository.com/artifact/androidx.asynclayoutinflater/asynclayoutinflater/1.0.0 dependency " +
+          "to your project to make use of asynchronous view inflation when adding view annotation!"
+      )
+    }
+  }
+
+  private val annotationMap = ConcurrentHashMap<String, ViewAnnotation>()
+  private val idLookupMap = HashMap<View, String>()
+
+  // structs needed for drawing, declare them only once
+  private val currentViewsDrawnMap = HashMap<String, ScreenCoordinate>()
+  private val positionDescriptorMap = HashMap<String, ViewPosition>()
+  private val idsToRepositionSet = HashSet<String>()
+  private val idsToAddSet = HashSet<String>()
+  private val idsToDeleteSet = HashSet<String>()
+
+  override fun addViewAnnotation(
+    @LayoutRes resId: Int,
+    options: ViewAnnotationOptions,
+    asyncInflateCallback: (View) -> Unit
+  ) {
+    validateOptions(options)
+    asyncInflater.inflate(resId, mapView) { view, _, _ ->
+      asyncInflateCallback.invoke(prepareViewAnnotation(view, options))
+    }
+  }
+
+  override fun addViewAnnotation(
+    @LayoutRes resId: Int,
+    options: ViewAnnotationOptions
+  ): View {
+    validateOptions(options)
+    val view = LayoutInflater.from(mapView.context).inflate(resId, mapView, false)
+    return prepareViewAnnotation(view, options)
+  }
+
+  override fun addViewAnnotation(view: View, options: ViewAnnotationOptions) {
+    idLookupMap[view]?.let {
+      throw RuntimeException(
+        "Trying to add view annotation that was already added before! " +
+          "Please consider deleting annotation view ($view) beforehand."
+      )
+    }
+    validateOptions(options)
+    prepareViewAnnotation(view, options)
+  }
+
+  override fun removeViewAnnotation(view: View): Boolean {
+    val id = idLookupMap[view] ?: return false
+    val annotation = annotationMap.remove(id) ?: return false
+    idLookupMap.remove(view)
+    view.viewTreeObserver.removeOnGlobalLayoutListener(annotation.globalLayoutListener)
+    mapView.removeView(view)
+    mapboxMap.removeViewAnnotation(id)
+    return true
+  }
+
+  override fun updateViewAnnotation(
+    view: View,
+    options: ViewAnnotationOptions,
+  ): Boolean {
+    val id = idLookupMap[view] ?: return false
+    annotationMap[id]?.let {
+      it.handleVisibility = options.visible == null
```

Comment:
IMO not very intuitive that `visible` if not set leaves view with automatic visibility control. 
Maybe `forceVisible` instead of plain `visible` will be more self-explanatory

Reply:
Yeah that confused me a bit to. @zmiao what do you think if we rename `visible` property in the options?
My proposal would be `forceHide` because the default behaviour is that the view is always visible if this is not set. (But then we would need to flip the logic in our implementations.)

Reply:
dropped comment about updating docs in gl-native's PR.
See not much sense in renaming as this property actually does exactly what it describes: changes visibility for view annotation core object. IMHO it's just better to provide better docs for this property as part of `ViewAnnotationOptions`.